### ISSUE TITLE
Logic for configuring the WordPress service appears to be missing password-propagation to MySQL

### ISSUE THREAD

**Issue by maratbn** [NixOS/nixpkgs#125318]
## Issue description

The logic for configuring the WordPress service appears to be missing the functionality to properly configure the new MySQL database for the WordPress instance when the option 'services.wordpress...database.createLocally' is configured as 'true'.

Specifically, the logic to configure the new MySQL database that the WordPress instance needs to be able to access seems to be missing the functionality to propagate its database password (specified via the option 'services.wordpress...database.passwordFile').

This results in the WordPress error screen "Error establishing a database connection" when visiting the WordPress instance with a web browser, as the MySQL service will reject the WordPress instance's attempt to access its database with the configured password.

The logic that propagates just the MySQL database user when 'createLocally' is 'true' is here -- https://github.com/NixOS/nixpkgs/blob/724ed08df02546fea2ab38613d615dd47461528c/nixos/modules/services/web-apps/wordpress.nix#L284

But there does not seem to be corresponding logic for the password.

 
### Steps to reproduce

(1) Specify the following configuration for 'services.wordpress' in 'nixos/configuration.nix':

  ...
  ...
  services.wordpress."mysite.mymachine" = {
    database = {
      host = "127.0.0.1";
      name = "wordpress";
      passwordFile = pkgs.writeText "wordpress-insecure-dbpass" "mypassword";
      createLocally = true;
    };
    virtualHost = {
      adminAddr = "admin@mymachine";
      serverAliases = [ "mysite.mymachine" ];
    };
  };
  ...
  ...

(2) Run command 'nixos-rebuild switch'.
(3) Open web browser and visit http://mysite.mymachine

   The screen "Error establishing a database connection" will appear.

(4) Start mysql console as root user with 'mysql'.
(5) Enter command 'show databases;'.  List of databases will appear.
(6) Enter command 'use mysql;'.  This will switch to the database MySQL uses to keep track of allowed "users" and their passwords.
(7) Enter command 'select User, Password from user;'.  Observe that the 'wordpress' "user" does not have a password.  (This configures MySQL to allow access to the Unix user 'wordpress' without a password; however, the 'httpd' instances started to run the WordPress site are running under a different Unix user 'wwwrun'.)

## Technical details

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.114, NixOS, 20.09.4064.1b688ca59ba (Nightingale)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.10`
 - channels(root): `"nixos-20.09.4064.1b688ca59ba"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

**Comment by aanderse** [comment]
The intention of `createLocally` is to only work with socket authentication. Why would you use a comparatively insecure authentication method (password) when you could use a secure one (socket authentication)?

**Comment by maratbn** [comment]
To explain, I am experimenting spinning up a WordPress site with a Nix configuration, and got confused by the 'passwordFile' option.

Appreciate the tip, understand that the correct configuration is actually this:

...
...
services.wordpress."mysite.mymachine" = {
  database = {
    host = "localhost";
    name = "wordpress";
    socket = "/run/mysqld/mysqld.sock";
    createLocally = true;
  };
  virtualHost = {
    adminAddr = "admin@mymachine";
    serverAliases = [ "mysite.mymachine" ];
  };
};
...
...

**Comment by mohe2015** [comment]
Close? @maratbn

**Comment by mohe2015** [comment]
Oh, nevermind there is an open PR

**Comment by maratbn** [comment]
Hi @mohe2015,

Sorry, I bumped into this issue several months ago, and submitted the simple PR https://github.com/NixOS/nixpkgs/pull/125329 that needs to be redone based on the feedback.

I also recall that when I started testing the socket authentication, it was not working each time in the way I expected, which led me to suspect that there may be another bug involved.  I wanted to understand what was happening first before submitting additional PRs, but unfortunately as I started investigating a bunch of high priority issues diverted my attention.

So still planning to investigate, but it may be a while.

As far as this bug report goes, understand that only socket authentication is supported, but if so, then why is there the `passwordFile` option?  If authentication via the password file is not supported then this option should be removed.  Or is password authentication still supported, but just not for creating a fresh database?  If so, what is the point of supporting password authentication in one mode but not in another?

**Comment by maratbn** [comment]
Hi @mohe2015,

Sorry, I bumped into this issue several months ago, and submitted the simple PR #125329 that needs to be redone based on the feedback.

I also recall that when I started testing the socket authentication, it was not working each time in the way I expected, which led me to suspect that there may be another bug involved. I wanted to understand what was happening first before submitting additional PRs, but unfortunately as I started investigating a bunch of high priority issues diverted my attention.

So still planning to investigate, but it may be a while.

For now closing this bug report as my main question has been answered and what I was specifically asking about does not appear to be a bug.
